File: perm.js
//Find all permutations of an array

function perm(arr) {
  if (arr.length === 1) return arr;
  let perm = [];
  arr.forEach(el => {
    if (perm.length > 0) {
      temp_perm = [];
      perm.forEach(p => {
        let temp = p;
        temp_perm.push([el].concat(temp));
        temp.forEach((item, index) => {
          temp_perm.push(
            temp
              .slice(0, index + 1)
              .concat([el])
              .concat(temp.slice(index + 1))
          );
        });
        // p.push(el);
        perm = temp_perm;
      });
    } else {
      perm.push([el]);
    }
  });
  console.log("Num of items: ", perm.length);
  return perm;
}

console.log(perm([4, 3, 2, 1]));

File: perm.rb


def perm ( arr)
    return arr if arr.length == 1
    res = []
    arr.each  do |el|
        if res.length == 0
            res << [el]
        else
            temp_res = [];
            p "res: ", res
            res.each do |e|
                e.each_index do |ind|
                    p "E: ", e, "Ind: ", ind
                  temp_res.push(e.slice(0,ind) + [el] + e.slice(ind, e.length - ind ))
                end
                temp_res.push(e+[el])
            end
            res = temp_res
        end    
    end
    res.map{|el| el.join("")}
end

p perm([1,2])
